truffleruby/src/main/ruby/core/truffle/cext.rb

```diff
@@ -23,6 +23,18 @@ def initialize(data)
 
   end
 
+  SYNC = Mutex.new
+
+  def self.execute_with_mutex(function, *args)
+    mine = Truffle::CExt::SYNC.owned?
+    Truffle::CExt::SYNC.lock unless mine
+    begin
+      Truffle::Interop.execute(function, *args)
+    ensure
+      Truffle::CExt::SYNC.unlock unless mine
+    end
+  end
```

Comment:
I think there's a `Mutex` method that takes a block and does the `ensure` for you.